### PR TITLE
Add Digital Twin Health Check

### PR DESCRIPTION
**Pull request by rafaarcangel**
**What this PR does / why we need it**:
This PR resolves the need to check Health for Azure Digital Twin.
Its offer two kinds of a health check:
- a liveness prove to test the connection to the resource
- a check for the DT existing models, comparing with a provided list of models.

See the README.md file.

**Which issue(s) this PR fixes**:
No one

Please reference the issue this PR will close: #_[issue number]_

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [X] Code compiles correctly
- [X] Created/updated tests
- [X] Unit tests passing
- [ ] End-to-end tests passing
- [X] Extended the documentation
- [X] Provided sample for the feature

### CODE REVIEW

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                if (!_managementClientConnections.TryGetValue(_clientId, out var managementClient))
+                {
+                    var credentials = new ClientSecretCredential(_tenantId, _clientId, _clientSecret);
+                    managementClient = new DigitalTwinsClient(new Uri(_hostName), credentials);
```
[the remark this answers is not in the document]
Comment (rafaarcangel) [reply src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs]: Updated names to digitalTwinClient

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            _models = models.Any() ? models : throw new ArgumentNullException(nameof(models));
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                if (!_managementClientConnections.TryGetValue(_clientId, out var managementClient))
+                {
+                    var credentials = new ClientSecretCredential(_tenantId, _clientId, _clientSecret);
```
[the remark this answers is not in the document]
Comment (rafaarcangel) [reply src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs]: This is outdated. The last version has two constructors, and one of them enables the user to send a TokenCredential.

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                        return new HealthCheckResult(context.Registration.FailureStatus, description: "No digital twin administration client connection can't be added into dictionary.");
+                    }
+                }
+
+                var response = managementClient.GetModelsAsync(cancellationToken: cancellationToken);
+                var models = new List<DigitalTwinsModelData>();
+                await foreach (var model in response) models.Add(model);
+
+                var unregistered = _models.Where(x => !models.Any(m => m.Id == x));
+                var unmapped = models.Where(x => !_models.Contains(x.Id)).Select(x => x.Id);
```
[the remark this answers is not in the document]
Comment (rafaarcangel) [reply src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs]: Removed unused 'unmapped' models

File: src/HealthChecks.AzureDigitalTwin/README.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            "MyDigitalTwinClientSecret",
+            "TenantId")
+}
+```
+
+---
+
+## _Digital Twin Model Health Check_
+
+This health check receibe a list of models id, and check if the Digital Twin has all models match with them.
```
[the remark this answers is not in the document]
Comment (rafaarcangel) [reply src/HealthChecks.AzureDigitalTwin/README.md]: Fixed

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinHealthCheck.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            _clientId = (!string.IsNullOrEmpty(clientId)) ? clientId : throw new ArgumentNullException(nameof(clientId));
+            _clientSecret = (!string.IsNullOrEmpty(clientSecret)) ? clientSecret : throw new ArgumentNullException(nameof(clientSecret));
+            _tenantId = (!string.IsNullOrEmpty(tenantId)) ? tenantId : throw new ArgumentNullException(nameof(tenantId));
+
+            _credentials = new AzureCredentialsFactory().FromServicePrincipal(_clientId, _clientSecret, _tenantId, AzureEnvironment.AzureGlobalCloud);
+        }
+
+        public AzureDigitalTwinHealthCheck(ServiceClientCredentials credentials)
+        {
+            _credentials = (credentials is not null) ? credentials : throw new ArgumentNullException(nameof(credentials));
```
Comment (sungam3r) [inline src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinHealthCheck.cs:35]:
```suggestion
            _credentials = credentials ?? throw new ArgumentNullException(nameof(credentials));
```

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinHealthCheck.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                if (!_managementClientConnections.TryGetValue(_clientId, out var managementClient))
```
Comment (sungam3r) [inline src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinHealthCheck.cs:42]: See https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/847 . The same problem may ne here.

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinInstanceHealthCheck.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+            _hostName = (!string.IsNullOrEmpty(hostName)) ? hostName : throw new ArgumentNullException(nameof(hostName));
+            _instanceName = (!string.IsNullOrEmpty(instanceName)) ? hostName : throw new ArgumentNullException(nameof(instanceName));
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                if (!_digitalTwinClientConnections.TryGetValue(_clientId, out var digitalTwinClient))
```
Comment (sungam3r) [inline src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinInstanceHealthCheck.cs:48]: and here

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            _tenantId = (!string.IsNullOrEmpty(tenantId)) ? tenantId : throw new ArgumentNullException(nameof(tenantId));
+            _credentials = new ClientSecretCredential(_tenantId, _clientId, _clientSecret);
+
+            _hostName = (!string.IsNullOrEmpty(hostName)) ? hostName : throw new ArgumentNullException(nameof(hostName));
+            _models = models.Any() ? models : throw new ArgumentNullException(nameof(models));
+        }
+
+        public AzureDigitalTwinModelsHealthCheck(TokenCredential credentials, string hostName, string[] models)
+        {
+            _credentials = (credentials is not null) ? credentials : throw new ArgumentNullException(nameof(credentials));
```
Comment (sungam3r) [inline src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs:40]:
```suggestion
            _credentials = credentials ?? throw new ArgumentNullException(nameof(credentials));
```

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+            _hostName = (!string.IsNullOrEmpty(hostName)) ? hostName : throw new ArgumentNullException(nameof(hostName));
+            _models = models.Any() ? models : throw new ArgumentNullException(nameof(models));
```
Comment (sungam3r) [inline src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs:43]: `ArgumentNullException` looks strange if one passes empty array.

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                if (!_digitalTwinClientConnections.TryGetValue(_clientId, out var digitalTwinClient))
```
Comment (sungam3r) [inline src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs:50]: and here

File: src/HealthChecks.AzureDigitalTwin/DependencyInjection/AzureDigitalTwinHealthCheckBuilderExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        /// <returns>The <see cref="IHealthChecksBuilder"/>.</returns>
+        public static IHealthChecksBuilder AddAzureDigitalTwin(this IHealthChecksBuilder builder, string clientId, string clientSecret, string tenantId, string name = default, HealthStatus? failureStatus = default, IEnumerable<string> tags = default, TimeSpan? timeout = default)
+        {
+            return builder.Add(new HealthCheckRegistration(
+                name ?? AZUREDIGITALTWIN_NAME,
+                sp => new AzureDigitalTwinHealthCheck(clientId, clientSecret, tenantId),
+                failureStatus,
+                tags,
+                timeout));
+        }
```
Comment (sungam3r) [inline src/HealthChecks.AzureDigitalTwin/DependencyInjection/AzureDigitalTwinHealthCheckBuilderExtensions.cs:40]:
```suggestion
        }
        
```

File: src/HealthChecks.AzureDigitalTwin/DependencyInjection/AzureDigitalTwinHealthCheckBuilderExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        /// <returns>The <see cref="IHealthChecksBuilder"/>.</returns>
+        public static IHealthChecksBuilder AddAzureDigitalTwinModels(this IHealthChecksBuilder builder, string clientId, string clientSecret, string tenantId, string hostName, string[] models, string name = default, HealthStatus? failureStatus = default, IEnumerable<string> tags = default, TimeSpan? timeout = default)
+        {
+            return builder.Add(new HealthCheckRegistration(
+                name ?? AZUREDIGITALTWINMODELS_NAME,
+                sp => new AzureDigitalTwinModelsHealthCheck(clientId, clientSecret, tenantId, hostName, models),
+                failureStatus,
+                tags,
+                timeout));
+        }
```
Comment (sungam3r) [inline src/HealthChecks.AzureDigitalTwin/DependencyInjection/AzureDigitalTwinHealthCheckBuilderExtensions.cs:91]:
```suggestion
        }
        
```

File: src/HealthChecks.AzureDigitalTwin/DependencyInjection/AzureDigitalTwinHealthCheckBuilderExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        /// <returns>The <see cref="IHealthChecksBuilder"/>.</returns>
+        public static IHealthChecksBuilder AddAzureDigitalTwinInstance(this IHealthChecksBuilder builder, string clientId, string clientSecret, string tenantId, string hostName, string instanceName, string name = default, HealthStatus? failureStatus = default, IEnumerable<string> tags = default, TimeSpan? timeout = default)
+        {
+            return builder.Add(new HealthCheckRegistration(
+                name ?? AZUREDIGITALTWININSTANCE_NAME,
+                sp => new AzureDigitalTwinInstanceHealthCheck(clientId, clientSecret, tenantId, hostName, instanceName),
+                failureStatus,
+                tags,
+                timeout));
+        }
```
Comment (sungam3r) [inline src/HealthChecks.AzureDigitalTwin/DependencyInjection/AzureDigitalTwinHealthCheckBuilderExtensions.cs:144]:
```suggestion
        }
        
```

File: src/HealthChecks.AzureDigitalTwin/README.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+  <br/>Default for model state if not specified is `azuredigitaltwininstance`.
+- `failureStatus`: The `HealthStatus` that should be reported when the health check fails. Default is `HealthStatus.Unhealthy`.
+- `tags`: A list of tags that can be used to filter sets of health checks.
+- `timeout`: A `System.TimeSpan` representing the timeout of the check.
+
+---
+
+## _Digital Twin Liveness Health Check_
+
+This health check provide the liveness status for the Azure Digital Twin resource client connection
```
Comment (sungam3r) [inline src/HealthChecks.AzureDigitalTwin/README.md:27]:
```suggestion
This health check provides the liveness status for the Azure Digital Twin resource client connection.
```

File: test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinInstanceUnitTests.cs [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+            var check = registration.Factory(serviceProvider);
+
+            registration.Name.Should().Be("azuredigitaltwininstance");
+            check.GetType().Should().Be(typeof(AzureDigitalTwinInstanceHealthCheck));
+
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinInstanceUnitTests.cs:35]:
```suggestion
```

File: test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinInstanceUnitTests.cs [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinInstanceUnitTests.cs:95]:
```suggestion
```

File: test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinInstanceUnitTests.cs [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+                .AddAzureDigitalTwinInstance(default(TokenCredential), string.Empty, string.Empty);
+
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+
+            Assert.Throws<ArgumentNullException>(() => registration.Factory(serviceProvider));
+        }
+
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinInstanceUnitTests.cs:133]:
```suggestion
```

File: test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinModelUnitTests.cs [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+            var check = registration.Factory(serviceProvider);
+
+            registration.Name.Should().Be("azuredigitaltwinmodels");
+            check.GetType().Should().Be(typeof(AzureDigitalTwinModelsHealthCheck));
+
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinModelUnitTests.cs:35]:
```suggestion
```

File: test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinModelUnitTests.cs [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+
+            Assert.Throws<ArgumentNullException>(() => registration.Factory(serviceProvider));
+        }
+
+
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinModelUnitTests.cs:76]:
```suggestion
```

File: test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinModelUnitTests.cs [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+            var check = registration.Factory(serviceProvider);
+
+            registration.Name.Should().Be("azuredigitaltwinmodels");
+            check.GetType().Should().Be(typeof(AzureDigitalTwinModelsHealthCheck));
+
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinModelUnitTests.cs:95]:
```suggestion
```

File: test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinUnitTests.cs [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+            var check = registration.Factory(serviceProvider);
+
+            registration.Name.Should().Be("azuredigitaltwin");
+            check.GetType().Should().Be(typeof(AzureDigitalTwinHealthCheck));
+
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinUnitTests.cs:30]:
```suggestion
```

File: test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinUnitTests.cs [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+
+            Assert.Throws<ArgumentNullException>(() => registration.Factory(serviceProvider));
+        }
+
+
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinUnitTests.cs:65]:
```suggestion
```

File: test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinUnitTests.cs [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+            var check = registration.Factory(serviceProvider);
+
+            registration.Name.Should().Be("azuredigitaltwin");
+            check.GetType().Should().Be(typeof(AzureDigitalTwinHealthCheck));
+
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/AzureDigitalTwin/AzureDigitalTwinUnitTests.cs:81]:
```suggestion
```

File: test/UnitTests/DependencyInjection/AzureDigitalTwin/Mocks/MockCredentials.cs [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            throw new NotImplementedException();
+        }
+
+        public override Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/AzureDigitalTwin/Mocks/MockCredentials.cs:35]:
```suggestion
```

**Comment by sungam3r** [comment]
@rafaarcangel Please rebase onto `master` if you are going to continue work on this PR.

**Comment by sungam3r** [comment]
Thanks. Please add CI/CD workflows to make new project testable. You may find many examples in `.github\worflows` folder. Also add 1-2 lines mentioning new package into README (nuget links, etc)

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinInstanceHealthCheck.cs (hunk 7 above)
Comment (rafaarcangel) [reply src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinInstanceHealthCheck.cs]: Updated!

File: src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs (hunk 9 above)
Comment (rafaarcangel) [reply src/HealthChecks.AzureDigitalTwin/AzureDigitalTwinModelsHealthCheck.cs]: For me it has no sense if a user try to compare/check for an empty DT but, it's possible that someone wants to be sure that has an empty initial models, so .... validation removed!

**Comment by rafaarcangel** [comment]
@sungam3r, you can check for all the changes :D

**Comment by carlosrecuero** [comment]
I will handle that.

**Comment by rafaarcangel** [comment]
Just updated those readme files

**Pull request merged**